src/engine/sparqlExpressions/AggregateExpression.h

```diff
@@ -113,49 +141,64 @@ class AggregateExpression : public SparqlExpression {
         return aggregateOperation._function(AD_FWD(x), AD_FWD(y), context);
       }
     };
-    // The operands *without* applying the `valueGetter`.
+
+    // A generator for the operands (before the value getter is applied to get
+    // the actual values).
     auto operands =
         makeGenerator(std::forward<Operand>(operand), inputSize, context);
+
+    // Set up cancellation handling.
     auto checkCancellation =
         [context](ad_utility::source_location location =
                       ad_utility::source_location::current()) {
           context->cancellationHandle_->throwIfCancelled(location);
         };
 
-    auto impl = [&valueGetter, context, &finalOperation, &callFunction,
-                 &checkCancellation](auto&& inputs) {
+    // Lambda to compute the aggregate of the given operands. This requires
+    // that `inputs` is not empty.
+    auto computeAggregate = [&valueGetter, context, &finalOperation,
+                             &aggregateTwoValues,
+                             &checkCancellation](auto&& inputs) {
       auto it = inputs.begin();
       AD_CORRECTNESS_CHECK(it != inputs.end());
 
-      using ResultType = std::decay_t<decltype(callFunction(
+      using ResultType = std::decay_t<decltype(aggregateTwoValues(
           std::move(valueGetter(*it, context)), valueGetter(*it, context)))>;
       ResultType result = valueGetter(*it, context);
       checkCancellation();
       size_t numValues = 1;
 
       for (++it; it != inputs.end(); ++it) {
-        result = callFunction(std::move(result),
-                              valueGetter(std::move(*it), context));
+        result = aggregateTwoValues(std::move(result),
+                                    valueGetter(std::move(*it), context));
         checkCancellation();
         ++numValues;
       }
       result = finalOperation(std::move(result), numValues);
       checkCancellation();
       return result;
     };
+
+    // Compute the aggregate (over all values or, if this is a DISTINCT
+    // aggregate, only over the distinct values).
+    //
+    // NOTE: If the GROUP BY is implicit and we have a single group, that
+    // group can be empty. Then we cannot start with the first value and
+    // successively aggregate the others. Instead, we have to return the
+    // "neutral element" of that aggregation operation.
```

Comment:
As I understand, this case has already been handled long before we reach this place, so this comment is a bit long for "Not that we have already handled the case of an empty input.